torch/distributed/_functional_collectives.py

```diff
@@ -625,8 +660,10 @@ def wrap(e: torch.Tensor):
         unwrapped_args = tree_map_only(AsyncCollectiveTensor, unwrap, args)
         unwrapped_kwargs = tree_map_only(AsyncCollectiveTensor, unwrap, kwargs)
 
-        # we don't wrap the result as it doesn't need to be waited on.
-        out = func(*unwrapped_args, **unwrapped_kwargs)
+        # we need this so that the grad_fn is correctly propagated
+        with torch.overrides.enable_reentrant_dispatch():
```

Comment:
what does this do?

Reply:
I don't really understand what it's doing under the hood but it makes the grad_fn be set correctly

Found it from this example with a tensor subclass + autograd https://github.com/pytorch/pytorch/blob/c8e117fb761235296a7eae6f88bce4ab68ab5c2a/torch/testing/_internal/logging_tensor.py#L94

It seems to do something with thread local storage but it's unclear why that is needed here https://github.com/pytorch/pytorch/blob/c8e117fb761235296a7eae6f88bce4ab68ab5c2a/aten/src/ATen/core/PythonFallbackKernel.h#L7

cc @albanD